Daniil Vasilev-17022A-official.contest.yandex.ru/contest/17022/run-report/30752651

SpeedOfMagic opened the pull request:
Отправляю задачу А с 3-го длинного контеста (17022) на ревью.
Ссылка на сабмит: official.contest.yandex.ru/contest/17022/run-report/30752651

SpeedOfMagic committed:
Create hash_map.h

ifsmirnov commented on hash_map.h:
```diff
@@ -0,0 +1,240 @@
+#include <utility>
+#include <vector>
+
+#pragma once
```
Принято в самом начале файла писать, до остальных инклудов.

ifsmirnov commented on hash_map.h:
```diff
@@ -0,0 +1,240 @@
+    vector<list<Node>> container;
+    list<Node*> pointers;
+    Hash hash;
+    size_t Size = 0, Capacity = 0;
```
Почему эти поля с большой буквы, а остальные с маленькой?

ifsmirnov commented on hash_map.h:
```diff
@@ -0,0 +1,240 @@
+
+ public:
+    explicit HashMap(const Hash& h = Hash()) : hash(h) {}
+    explicit HashMap(const HashMap& hm) : hash(hm.hash_function()) {
```
1) не надо сокращать, лучше const HashMap& hashMap или const HashMap& other.
2) зачем конструктор копирования explicit?

ifsmirnov commented on hash_map.h:
```diff
@@ -0,0 +1,240 @@
+            insert(i);
+    }
+
+    const HashMap& operator=(const HashMap& hm) {
```
Оператор присваивания должен возвращать неконстантную ссылку.

ifsmirnov commented on hash_map.h:
```diff
@@ -0,0 +1,240 @@
+    }
+
+    const HashMap& operator=(const HashMap& hm) {
+        if (container == hm.container)
```
Скорее всего имелось в виду &container == &hm.container, текущая проверка довольно странная. Вообще обычно пишут if (this != &other).

ifsmirnov commented on hash_map.h:
```diff
@@ -0,0 +1,240 @@
+    bool empty() const noexcept { return size() == 0; }
+    Hash hash_function() const noexcept { return hash; }
+
+    size_t h(const KeyType& key) const noexcept {
```
Название ни о чём не говорит.

Откуда предположение, что вызов hash(key) не бросает исключения?

ifsmirnov commented on hash_map.h:
```diff
@@ -0,0 +1,240 @@
+    }
+
+    void insert(const pair<const KeyType, ValueType>& p) {
+        if (find(p.first) == end()) {
```
Тут лучше бы смотрелся early exit (инвертировать условие и сделать return).

ifsmirnov commented on hash_map.h:
```diff
@@ -0,0 +1,240 @@
+                clear();
+                Capacity = Capacity ? (Capacity << 1) : 1;
+                container.resize(Capacity);
+                for (const auto& i : allPairs)
```
i обычно используется для индекса, здесь лучше назвать переменную pair, например. Выше тоже.

ifsmirnov commented on hash_map.h:
```diff
@@ -0,0 +1,240 @@
+        typename list<Node*>::iterator it;
+        iterator() = default;
+        explicit iterator(
+                const typename list<Node*>::iterator& i) : it(i) {}
```
Зачем тут перенос строки? Одна строка влезает в 78 символов, что удовлетворяет даже самым жёстким кодстайлам.

ifsmirnov commented on hash_map.h:
```diff
@@ -0,0 +1,240 @@
+class HashMap {
+ public:
+    struct Node {
+        pair<const KeyType, ValueType> p;
```
Однобуквенные переменые не ок.

ifsmirnov commented on hash_map.h:
```diff
@@ -0,0 +1,240 @@
+        pair<const KeyType, ValueType> p;
+        typename list<Node*>::iterator ptr;
+
+        explicit Node(const pair<const KeyType, ValueType>& _p) : p(_p) {}
```
В конструкторе название переменной может совпадать с названием поля:

```
explicit Node(const std::pair<const KeyType, ValueType>& p) : p(p) {}
```

SpeedOfMagic replied:
Флаг компиляции -Wshadow жалуется, что название поля совпадает с названием переменной.

ifsmirnov commented on hash_map.h:
```diff
@@ -0,0 +1,240 @@
+        }
+
+        pair<const KeyType, ValueType>& operator*() const {
+            return (**it).p;
```
К слову об однобуквенных переменных -- я смотрю на эту строчку и плохо понимаю, что такое p.

ifsmirnov commented on hash_map.h:
```diff
@@ -0,0 +1,240 @@
+            return end();
+        size_t cell = h(key);
+        for (auto i = container[cell].begin(); i != container[cell].end(); ++i)
+            if ((*i).p.first == key)
```
i->p.first

ifsmirnov commented on hash_map.h:
```diff
@@ -0,0 +1,240 @@
+    void erase(const KeyType& key) {
+        if (Capacity == 0)
+            return;
+        size_t c = h(key);
```
cellIlndex

ifsmirnov reviewed (changes requested):
По стилю -- стоит избавиться от однобуквенных переменных (всегда лучше написать bucketIndex вместо c и getBucketIndex вместо h), навести порядок в переводах строк между функциями (где-то есть, где-то нет), и ещё по мелочи я отдельно комментировал.

По логике -- стоит попробовать избавиться от дублирования кода в find. Сейчас это две идентичные функции, определённо можно выделить общую часть.

SpeedOfMagic committed:
Resolved requested changes in hash_map.h

Ссылка на сабмит: official.contest.yandex.ru/contest/17022/run-report/30975626/
Про функцию find: т.к. она теперь занимает 6 строк, то мне кажется, что выделение какой-то общей части займёт ещё больше строк чем сейчас.

SpeedOfMagic committed:
Rewrote find function

Ссылка на сабмит:  official.contest.yandex.ru/contest/17022/run-report/30981320/
Вроде переделал find, но в 206 строке использую довольно сомнительный подход. Если тебе не нравится, то подскажи как можно сделать лучше.

SpeedOfMagic committed:
Made non-const find function more clear

Вынес перевод константного итератора в неконстантный в отдельную функцию, переделал неконстантную функцию find, чтобы она использовала эту функцию.
